dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -389,7 +381,11 @@ parsers embedded = Parsers {..}
                 b <- importExpression_
                 return (sep, b)
 
-            return (foldl' app (f a) bs)
+            let c = foldl' app (f a) bs
+
+            let type_ = if eqExprBySrc a c then ImportExpr else ApplicationExpr
```

Comment:
Do you need to check the `Src` to know if they are equal?  It seems like you could determine that more reliably by checking if `maybeMessage == Nothing && null bs`